theano/gpuarray/basic_ops.py

```diff
@@ -1377,15 +1377,140 @@ class GpuSplit(HideC, Split):
+        size_t len_along_axis, sum_of_splits = 0, current_split_start = 0;
+        %(splits_dtype)s current_split_length = 0;
+        GpuArray view;
+        int i;
+
+        /* Check inputs. */
+
+        if (splits_count != %(expected_splits_count)s) {
+            PyErr_Format(PyExc_ValueError,
+                "GpuSplit: splits count (%%d) != expected count (%%d).", splits_count, %(expected_splits_count)s);
+            %(fail)s
+        }
+        if (axis < 0) {
+            axis += ndim;
+        }
+        if (axis < 0 || axis >= ndim) {
+            PyErr_Format(PyExc_IndexError, "GpuSplit: invalid axis %%d for a %%d-D array.", axis, ndim);
+            %(fail)s
+        }
+        len_along_axis = PyGpuArray_DIM(%(x)s, axis);
+        for (i = 0; i < splits_count; ++i) {
+            current_split_length = *(%(splits_dtype)s*)PyArray_GETPTR1(%(splits)s, i);
+            if (current_split_length < 0) {
+                PyErr_Format(PyExc_ValueError,
+                    "GpuSplit: you try to take a negative number (%%ld) of elements.", current_split_length);
+                %(fail)s
+            }
+            sum_of_splits += current_split_length;
+        }
+        if (sum_of_splits != len_along_axis) {
+            PyErr_Format(PyExc_ValueError, "GpuSplit: the splits sums to %%ld, expected %%ld.", sum_of_splits, len_along_axis);
+            %(fail)s
+        }
+
+        /* Check outputs. */
+        %(full_code_for_checking_outputs)s
+
+        /* Compute splits. */
+        if (GpuArray_view(&view, &%(x)s->ga) != GA_NO_ERROR) {
+            PyErr_SetString(PyExc_RuntimeError, "GpuSplit: unable to create a view of the input.");
+            %(fail)s
+        }
+        %(full_code_for_splitting)s
+
+        /* Free memory. */
+        GpuArray_clear(&view);
+
+        /* Code added if synchronization is enabled. */
+        %(full_code_if_sync)s
+        """
+
+        for split_index, output in enumerate(outputs):
+            # When checking output, we allocate a PyGpuArrayObject with 0 dims
+            # (that is, an object with as few memory as possible), as its GpuArray field
+            # will be cleared and re-used as a view during split operations.
+            codes_for_checking_outputs.append("""
+            if (theano_prep_output(&%(output)s, 0, NULL, %(x_typecode)s, x_order, %(x)s->context) != 0) {
+                PyErr_SetString(PyExc_RuntimeError, "GpuSplit: unable to prepare an output.");
+                %(fail)s
+            }
+            """ % locals())
+
+            code_for_splitting.append("""
+            current_split_length = * (%(splits_dtype)s*) PyArray_GETPTR1(%(splits)s, %(split_index)s);
+            view.offset = PyGpuArray_STRIDE(%(x)s, axis) * current_split_start;
+            view.dimensions[axis] = current_split_length;
+            GpuArray_fix_flags(&view);
+            GpuArray_clear(&%(output)s->ga);
+            if (GpuArray_view(&%(output)s->ga, &view) != GA_NO_ERROR) {
```

Comment:
This part is almost as costly as doing Py_XDECREF() followed by pygpu_view().

It also introduces the only op in all of the new backend for which the gpudata pointer can change even if the python object didn't.  Scan relies on the new backend not doing this, so it will break in interesting ways (as in not fun to debug silent wrong data).

Also if the goal here is to reuse output memory then it fails, because this frees the output memory anyway.

So this code manages to
a) be significantly more complex,
b) not do anything inplace,
c) have some cases wrong and
d) break scan.

For all of these reasons I am vehemently against merging.